DBZ-8827 Documentation for Debezium Platform (#6263)
debezium/debezium: documentation/modules/ROOT/pages/operations/debezium-platform.adoc

[id="debezium-platform"]
= {prodname} Management Platform

:linkattrs:
:icons: font
:toc:
:toclevels: 3
:toc-placement: macro

toc::[]

ifdef::community[]
[NOTE]
====
This project is currently in an incubating state.
The exact semantics, configuration options, and so forth are subject to change, based on the feedback that we receive.
====
endif::community[]

{prodname} Management Platform aims to simplify the deployment of {prodname} to various environments in a highly opinionated manner.
To achieve this goal, the platform uses a data-centric view of {prodname} components.

It is for us a normal evolution from {prodname} Server. We have provided the operator - to easily operate it in Kubernetes environments - and now we are providing a high-level abstraction to deploy your data pipelines in different environments leveraging on {prodname} Server.

== Basic concepts
In {prodname} Management Platform there are four main concepts:

Source:: Defines the source of your data
Destination:: Defines the destination of your data
Transform:: Define how a single data is transformed

> Pipeline:: Defines how your data flows from a source to a destination while being transformed along the way.

Reviewer: @roldanbob in the DMP (Debezium Management Platform), we decompose the configuration for Debezium Server into reusable parts. 

So what you would normally see defined in `debezium.transforms`, we take each of those and create a standalone component we call `Transform` that defines the transform type and its entire configuration set.  So when the user creates a pipeline, which is the glue of combining a source, sink, and 0 or more transforms, a user can define the transform once and reuse it in multiple pipelines.

> Pipeline:: Defines how your data flows from a source to a destination while being transformed along the way.

Author: @roldanbob what do you think of something like 

`Transform:: Defines how a single data event is transformed as they flow through the pipeline.`

> Pipeline:: Defines how your data flows from a source to a destination while being transformed along the way.

Reviewer 2: I think that because we're talking about an event, this should read _"Defines how a single data event is transformed as **it** flows through the pipeline."_

Once you define a pipeline, how it will be deployed is in charge of the platform.

Every pipeline will be mapped to a Debezium Server instance. For the Kubernetes environment (the only supported as of now) this will correspond to a DebeziumServer custom resource.

=== Architecture
The platform is composed of two main components:

Conductor:: The back-end component which provides a set of APIs to orchestrate and control {prodname} deployments.

> Stage:: The front-end component which provides a user interface to interact with the Conductor.

Reviewer: ```suggestion
Stage:: The front-end component that provides a user interface to interact with the Conductor.
```

[source,bash]
----
helm repo add debezium https://charts.debezium.io
----

after that, you can install a specific version with the following command:

[source, bash]
----
helm install debezium-platform debezium/debezium-platform --version 3.1.0-beta1 --set database.enabled=true --set domain.url=platform.debezium.io
----

as an alternative to this, you can also install it via OCI artifact with the following command:

[source, bash]
----
helm install debezium-platform --set database.enabled=true --set domain.url=platform.debezium.io --version 3.1.0-beta1 oci://quay.io/debezium-charts/debezium-platform
----

the `domain.url` is the only required property, and it is used as `host` in the `Ingress` definition.
The `database.enabled` properties has been added for simplicity since in that way the PostgreSQL database, required by the conductor service, is automatically deployed.
This is discouraged in production environment, so in that case please do not use it and provide instead your database instance.

The following tables lists all the chart's properties:

[cols="1,3,1", options="header"]
|===
|Name |Description |Default

|domain.url
|domain used as ingress host
|""

|stage.image

> |Image for the stage (UI)

Author: Effectively this images is used for the `stage` pod and not for build.

> |Image for the stage (UI)

Author: Well "stage" is the name of the pod that helm will deploy on K8s using this image.

|quay.io/debezium/platform-stage:<release_tag>

|conductor.image

> |Image for the conductor

Author: Yes, it is.

|quay.io/debezium/platform-conductor:<release_tag>

|conductor.offset.existingConfigMap
|Name of the config map used to store conductor offsets. If empty it will be automatically created.
|""

|database.enabled
|Enable the installation of PostgreSQL by the chart
|false

|database.name
|Name of the database used by the platform
|postgres

|database.host
|Host of the database used by the platform
|postgres

|database.auth.existingSecret

> |Name of the secret where `username` and `password` - of the database used by the platform - are stored . If empty a secret will be created using the `username` and `password` properties.

Reviewer: I believe what @mfvitale is attempting to convey is that if you supply `database.auth.existingSecret`, the username and value are sourced from the secret. However, if this field is left empty and no secret is supplied, the values of `database.auth.username` and `database.auth.password` are used.

> |Name of the secret where `username` and `password` - of the database used by the platform - are stored . If empty a secret will be created using the `username` and `password` properties.

Reviewer 2: Ok, that's what I thought was going on, but the wording here being different than what appears later made me question my understanding. Thanks.

When this is used, you don't need to set `database.auth.username` and `database.auth.password`.
|""

|database.auth.username
|Username of the database used by the platform
|user

|database.auth.password
|Password of the database used by the platform
|password

|offset.reusePlatformDatabase
|Pipelines will use database to store offsets. By default, the database used by the platform is used.
If you want to use a dedicated one set this property to false
|true

|offset.database.name
|Name of the database used by the platform for storing the offsets
|postgres

|offset.database.host
|Host of the database used by the platform for storing the offsets
|postgres

|offset.database.port
|Port of the database used by the platform for storing the offsets
|5432

|offset.database.auth.existingSecret
|Name of the secret where `username` and `password` - of the database used by the platform for storing the offsets - are stored. If not set `offset.database.auth.username` and `offset.database.auth.password` will be used.

> When this is used, you don't need to set `offset.database.auth.username` and `offset.database.auth.password`.

Reviewer 2: ```suggestion
If you provide the name of a secret, do not set the `offset.database.auth.username` and `offset.database.auth.password` properties.
```

|""

|offset.database.auth.username

> |Username of the database used by the platform for storing the offsets

Reviewer 2: ```suggestion
|Username through which the platform connects to the offsets database. 
```

|user

|offset.database.auth.password

> |Password of the database used by the platform for storing the offsets

Reviewer 2: ```suggestion
|Password for the offsets database user specified by `offset.database.auth.username`.

```

|password

|schemaHistory.reusePlatformDatabase

> |Pipelines will use database to store schema history. By default, the database used by the conductor service is used. If you want to use a dedicated one set this property to false

Reviewer 2: ```suggestion
|Specifies whether pipelines use the configured platform database to store the schema history. 
To configure pipelines to use a different, dedicated database to store the schema history, set the value to `false`.
```

|true

|schemaHistory.database.name

> |Name of the database used by the platform for storing the schema history

Reviewer 2: ```suggestion
|Name of the dedicated database where the platform stores the schema history.
```

|postgres

|schemaHistory.database.host

> |Host of the database used by the platform for storing the schema history

Reviewer 2: ```suggestion
|Host for the dedicated database where the platform stores the schema history.
```

|postgres

|schemaHistory.database.port

> |Port of the database used by the platform for storing the schema history

Reviewer 2: ```suggestion
|Port through which the platform connects to the dedicated database where it stores the schema history.
```

|5432

|schemaHistory.database.auth.existingSecret

> |Name of the secret where `username` and `password` - of the database used by the platform for storing the schema history -are stored. If not set `schemaHistory.database.auth.username` and `schemaHistory.database.auth.password` will be used.

Reviewer 2: ```suggestion
|Name of the secret that stores the `username` and `password` that the platform uses to authenticate with the database that stores the schema history. 
If you do not specify value, instead of using a secret to store credentials, the platform uses the values of the schemaHistory.database.auth.username` and `schemaHistory.database.auth.password` properties to authenticate with the database.
```

> When this is used, you don't need to set `schemaHistory.database.auth.username` and `schemaHistory.database.auth.password`.

Reviewer 2: ```suggestion
If you provide the name of a secret, do not set the `schemaHistory.database.auth.username` and `schemaHistory.database.auth.password` properties.
```

|""

|schemaHistory.database.auth.username

> |Username of the database used by the platform for storing the schema history

Reviewer 2: ```suggestion
|Username through which the platform connects to the schema history database.
```

|user

|schemaHistory.database.auth.password

> |Password of the database used by the platform for storing the schema history

Reviewer 2: ```suggestion
|Password for the schema history database user specified by `schemaHistory.database.auth.username` property.
```

|password

|env

> |List of env variable to pass to the conductor

Reviewer 2: ```suggestion
|List of environment variables to pass to the conductor.
```

==== Edit a source
To edit a source, go to the `Source` menu and then click the `action` menu of the source you want to edit, then click `Edit`.

[NOTE]
====
Editing a source will affect all pipelines that use it.
====

=== Destination
In this section, you can define the destinations where your source data will be sent.
All {prodname} Server sinks are available as destination.
When you create a destination, it can be shared between different pipelines, which means that every change to a destination will be reflected in every pipeline that uses it.

==== Create a new destination
In this section, you can configure your destination in two different ways. You can use the `Form Editor`, where you can enter the name of the destination and a description, and then specify the list of properties for the specific destination system.
Refer to the {prodname} sink-specific documentation page for the available properties.

==== Configure the destination with Smart Editor
The other option is the Smart Editor, where you can directly edit/paste the JSON configuration.
For those familiar with {prodname}, this is quite similar to {prodname} Server `sink` configuration section with small differences.

> Usually you have that the configuration of a particular sink are prefixed with `debezium.sink.<sink_name>` where `<sink_name` is the sink `type`.

Reviewer: ```suggestion
Usually you have that the configuration of a particular sink are prefixed with `debezium.sink.<sink_name>` where `<sink_name>` is the sink `type`.
```

When a transform as a predicate configured, you will see a image:predicate-icon.png[alt text] on top of it. A tooltip will show the name of the predicate used.

Once finished designing your pipeline you can click on `Configure Pipeline` and then you can now configure the name, the description and the logging level.

==== Delete
To delete a pipeline, go to the `Pipeline` menu and then click the `action` menu of the pipeline you want to delete, then click `Delete`.
Only the pipeline will be removed, the source, the destination and the transformations will not be deleted.

==== Edit
To edit a pipeline, go to the `Pipeline` menu and then click the `action` menu of the destination you want to edit, then click `Edit pipeline`.
As first step you can modify the transformations through the pipeline designer, we will go deeper in this part in the next section, and then you can edit the name, the description and the log level.

===== Remove or ordering transformations
Once you are in the `Pipeline designer` you can modify the order of transformations or delete one by clicking on the image:transformation-box-edit.png[alt text] icon.

==== Observability

> Currently, th observability is limited to logs that can be accessed from two places.

Reviewer: ```suggestion
Observability is currently limited to viewing the Debezium Server logs.
```

Instead, specify an existing database instance, by setting the `database.name`, `database.host`, and other properties required to connect to the database. 
See the following table for more information.
====

The following tables lists all the chart's properties:

[cols="1,3,1", options="header"]
|===
|Name |Description |Default

|domain.url
|Domain used as the ingress host
|""

|stage.image
|Image for the stage (UI)
|quay.io/debezium/platform-stage:<release_tag>

|conductor.image
|Image for the conductor
|quay.io/debezium/platform-conductor:<release_tag>

|conductor.offset.existingConfigMap
|Name of the ConfigMap that stores conductor offsets. 
If no value is specified, Helm creates a ConfigMap automatically. 
|""

|database.enabled
|Enables Helm to install PostgreSQL.
|false

|database.name
|Name of an existing database where you want the platform to store data.
|postgres

|database.host
|Host of the database that you want the platform to use.
|postgres

|database.auth.existingSecret
|Name of the secret that stores the `username` and `password` that the platform uses to authenticate with the database. 

> If no value is specified, a secret is created using the `username` and `password` properties.

Author: Well the subject that will auto create the secret in this case is helm. So replacing `platform` with `helm` should be fine.

  }
}
----

You just need to copy the `config` section, removing the `connector.class`, since it is already provided with the `type`.
In the future we will eventually support the Kafka Connect and/or Debezium Server format directly.

The final `json` should something like the following
[source,json,options="nowrap"]
----
{
    "name": "my-source",
    "description": "This is my first source",
    "type": "io.debezium.connector.mysql.MySqlConnector",
    "schema": "schema123",
    "vaults": [],
    "config": {
        "database.hostname": "mysql",
        "database.port": "3306",
        "database.user": "debezium",
        "database.password": "dbz",
        "database.server.id": "184054",
        "topic.prefix": "dbserver1",
        "database.include.list": "inventory"
    }
}
----

==== Delete a source
To delete a source, go to the `Source` menu and then click the `action` menu of the source you want to delete, then click `Delete`.
A source can be deleted only if it is not used in any pipeline; otherwise, you will receive an error.
When the source is no longer used in any pipeline, you can delete it using the `Delete` option.

==== Edit a source

> To edit a source, go to the `Source` menu and then click the `action` menu of the source you want to edit, then click `Edit`.

Author: No, fixed it. Good catch.

==== Create
In the pipeline menu you can click on `Create your first pipeline` and you will get into the `Pipeline Designer`.
Here you can add the pieces that composes you data pipeline. First of all, you need to add a source clicking on the `+ Source` box and the you can either choose a previously created source or directly create a new one.

Similarly, you can add a destination clicking on the `+ Destination` box.

If you want to apply some transformation to your data, you can add it in the same way just clicking on the `+ Transform` box.

When a transform as a predicate configured, you will see a image:predicate-icon.png[alt text] on top of it. A tooltip will show the name of the predicate used.

Once finished designing your pipeline you can click on `Configure Pipeline` and then you can now configure the name, the description and the logging level.

==== Delete
To delete a pipeline, go to the `Pipeline` menu and then click the `action` menu of the pipeline you want to delete, then click `Delete`.
Only the pipeline will be removed, the source, the destination and the transformations will not be deleted.

==== Edit

> To edit a pipeline, go to the `Pipeline` menu and then click the `action` menu of the destination you want to edit, then click `Edit pipeline`.

Reviewer 2: ```suggestion
. From the platform UI, open the **Pipeline** menu, click the **action** menu for the pipeline that you want to edit, and then click **Edit pipeline**.
```

> As first step you can modify the transformations through the pipeline designer, we will go deeper in this part in the next section, and then you can edit the name, the description and the log level.

Reviewer 2: ```suggestion
. Use the pipeline designer to modify transformations as needed.
You can edit the name, description, and log level for the transformation.
For more information about using the pipeline designer, see xref:debezium-platform-using-the-pipeline-designer-to-remove-and-order-transformations[Using the Pipeline designer to remove and order transformations].
```

===== Remove or ordering transformations

> Once you are in the `Pipeline designer` you can modify the order of transformations or delete one by clicking on the image:transformation-box-edit.png[alt text] icon.

Reviewer 2: ```suggestion
From the `Pipeline designer`, you can modify the order of transformations or delete transformations.

* To delete a transformation, click the Transform box (image:transformation-box-edit.png[transformation box]).
```
... and click **Delete**? Or is the deletion done simply by clicking on the box?
And how do you arrange transformation? Is drag-and-drop enabled in the UI?

Author: I'm referring to this 

![image](https://github.com/user-attachments/assets/9af49447-28fa-400b-9950-f736d87a443c)

you click on the "pencil" icon and you will go on the edit

![image](https://github.com/user-attachments/assets/7e9c3eb5-4780-44c4-a5c2-a029729a0457)

==== Observability

> Observability is currently limited to viewing the Debezium Server logs.

Reviewer 2: ```suggestion
The {prodname} platform provides easy access to the {prodname} Server logs.
```

> You can go to `Pipeline` then click on the pipeline name of your interest and then click on the `Pipeline logs`.

Reviewer 2: ```suggestion
. From the platform UI, click **Pipeline**, click the name of the pipeline you want to monitor, and then click **Pipeline logs**.
```

> The other ways is to go to `Pipeline` and then click the `action` menu of the pipeline you are interested, then click `View logs`.

Reviewer 2: ```suggestion

```
I'd suggest removing the alternative method on the principle that it's better for the documentation to be opinionated, and leave it to users to discover shortcuts and alternative methods that are available through mechanisms that we can assume the user is familiar with.
